extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/AggregationResultUpdater.java

```diff
@@ -0,0 +1,583 @@
+import org.openrdf.query.algebra.MathExpr.MathOp;
+import org.openrdf.query.algebra.evaluation.ValueExprEvaluationException;
+import org.openrdf.query.algebra.evaluation.util.MathUtil;
+import org.openrdf.query.algebra.evaluation.util.ValueComparator;
+import org.openrdf.query.impl.MapBindingSet;
+
+import com.google.common.collect.ImmutableMap;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Updates the results of an Aggregate node when its child has added a new Binding Set to its results.
+ */
+@DefaultAnnotation(NonNull.class)
+public class AggregationResultUpdater {
+    private static final Logger log = Logger.getLogger(AggregationResultUpdater.class);
+
+    private static final AggregationStateSerDe AGG_STATE_SERDE = new ObjectSerializationAggregationStateSerDe();
+    private static final VisibilityBindingSetSerde BS_SERDE = new VisibilityBindingSetSerde();
+
+    private static final ImmutableMap<AggregationType, AggregationFunction> FUNCTIONS;
+    static {
+        final ImmutableMap.Builder<AggregationType, AggregationFunction> builder = ImmutableMap.builder();
+        builder.put(AggregationType.COUNT, new CountFunction());
+        builder.put(AggregationType.SUM, new SumFunction());
+        builder.put(AggregationType.AVERAGE, new AverageFunction());
+        builder.put(AggregationType.MIN, new MinFunction());
+        builder.put(AggregationType.MAX, new MaxFunction());
+        FUNCTIONS = builder.build();
+    }
+
+    /**
+     * Updates the results of an Aggregation node where its child has emitted a new Binding Set.
+     *
+     * @param tx - The transaction all Fluo queries will use. (not null)
+     * @oaram childRow - The Row Key of the child Binding Set that changed. (not null)
+     * @param childCol - The Column Key of the child Binding Set that changed. (not null)
+     * @param aggregationMetadata - The metadata of the Aggregation node whose results will be updated. (not null)
+     * @throws Exception The update could not be successfulyl performed.
+     */
+    public void updateAggregateResults(
+            final TransactionBase tx,
+            final Bytes childRow,
+            final Column childCol,
+            final AggregationMetadata aggregationMetadata) throws Exception {
+        requireNonNull(tx);
+        requireNonNull(childRow);
+        requireNonNull(childCol);
+        requireNonNull(aggregationMetadata);
+
+        // Make sure the child binding set has not already been handled.
+        final Bytes childValue = tx.get(childRow, childCol);
+        final VisibilityBindingSet childBindingSet = BS_SERDE.deserialize(childValue);
+
+        log.trace(
+                "Transaction ID: " + tx.getStartTimestamp() + "\n" +
+                "Row Key: " + childRow + "\n" +
+                "Column Key: " + childCol + "\n" +
+                "Binding Set:\n" + childBindingSet + "\n");
```

Comment:
Done. You're still going to need to do a bit of surgery in the future if you want to generalize this problem. I think the Row and Column being passed in make more sense than the observation. The nodes will also have to fetch their own metadata.